app-server/database.js

```diff
@@ -0,0 +1,771 @@
+       Persons when a person is attempting to establish an account and where
+       they have already been referred to in other data (being cast in a role
+       in a production that is known to the system, for example). In order to
+       claim an account, the user must provide some form of contact information
+       so that an admin is able to reach out to them in the case of invalid
+       claiming of an account. All claimed Persons will be marked as such in
+       order to aid in tracking down issues related to improper claiming of
+       Persons.
+    3. Users have three levels: Global Admin, Org Admin, and Participant.
+       Technically, there is some conceptual muddiness with Participants.
+       Participant includes users who participate in *creating* theatre events
+       as well as audience members who merely are engaged in *attending* theatre
+       events. Anyone can attend events, obviously. But the way the permissions
+       are modelled, anyone could also begin to participate in creating events.
+       All that would change is that the user would assign or be assigned to
+       one or more collaborator roles.
+ */
+
+class User {
+
+  /* 
+   * API methods:
+   */
+
+  static getById(id) {
+    return new Promise((resolve, reject) => {
+      var connection = getMysqlConnection();
+      var query = `select u.*, up.access_level from user u, user_permissions up
+                   where u.id = ? and u.id = up.user_id`;
+      connection.query(query, [id], (err, results) => {
+        if (err) {
+          reject(err);
+        } else {
+          if (results.length) {
+            var user = User._userRecordToObject(results[0]);
+            resolve(user);
+          } else {
+            resolve(null);
+          }
+        }
+        connection.destroy();
+      });
+    });
+  }
+  static getUserAndPasswordByEmail(email) {
+    return new Promise((resolve, reject) => {
+      var connection = getMysqlConnection();
+      connection.query("select * from user where lower(email) = ?", [email.toLowerCase()], 
+        (err, results) => {
+          if (err) {
+            reject(err);
+          } else {
+            if (results.length) {
+              var user = User._userRecordToObject(results[0]);
+              var password = results[0].password;
+              resolve({user, password});
+            } else {
+              resolve({});
+            }
+          }
+          connection.destroy();
+        });
+    });
+  }
+  static create(inputUser) {
+    return new Promise((resolve, reject) => {
+      var {email, password} = inputUser;
+
+      var insertQuery = "insert into user (email, password) values (?, ?)";
+      var addPermsQuery = "insert into user_permissions (user_id, access_level) values (?, 'PARTICIPANT')";
+
+      var connection = getMysqlConnection();
+      connection.query(insertQuery, [email, password], (insertErr, insertResult) => {
```

Comment:
With the helpers I mentioned, the nesting can be reduced:

```javascript
static create(inputUser) {
  // Do as much of the non-asynchronous work here. That way you fail fast if there are errors.
  var {email, password} = inputUser;
  var connection = getMysqlConnection();
  var insertQuery = "...";
  // Always return the promise, even if you don't need to use it yet.
  return queryConnection(connection, insertQuery, [email, password])
    .then(/* some common insert result helper */)
    .then((id) => {
      var permsQuery = "...";
      return queryConnection(connection, perms, [id]);
    })
    // You could also combine the query and helper into a specific insert helper.
    .then(/* result helper */)
    // I imagine this helper looks up a created object based on the insertId and type.
    .then(id => lookUp("user", id)))
    // Easier than handling it specifically every place it matters.
    .finally(() => connection.destroy());
}
```

I think it's a good example of not having to track all your resolve/rejects. It means you can describe the flow in the promise chain, which I find a lot more readable and maintainable.